### PR TITLE
passing the bucket into a loop

### PR DESCRIPTION
**Pull request by remicousin**
For now, ignoring the possibilities of multiple years/growing-seasons.

Turns out that I forgot to clip the wb to 0 in the bucket step. wb can not be negative. I guess it would be more logical to clip et so that it can't be greater than sm(t-1) + peff(t) but that is not how Eunjin does it. She also sets wb to 0 if negative.

Also corrected some typos or practicalities.

After this PR, we can start build up the loop. So next should be Kc this time for real.

### CODE REVIEW

File: onset_maproom/agronomy.py [hunk 1]
```diff
@@ -1,10 +1,12 @@
 import xarray as xr
+import numpy as np
+
 
 def soil_plant_water_bucket(
```
Comment (aaron-kaplan) [inline onset_maproom/agronomy.py:5]: Find a better name? Maybe something with the word "step" in it?

"Bucket" is kind of cute, because buckets hold water, but kind of confusing for the same reason.

File: onset_maproom/agronomy.py [hunk 2]
```diff
@@ -29,19 +31,84 @@ def soil_plant_water_bucket(
...
     sm = wb - drainage
+    return sm, drainage
+
+
+def soil_plant_water_balance(
+    peffective,
+    et,
+    taw,
+    sminit,
+    dim="T",
```
Comment (aaron-kaplan) [inline onset_maproom/agronomy.py:61]: call it `time_dim`?

File: onset_maproom/agronomy.py [hunk 3]
```diff
@@ -29,19 +31,84 @@ def soil_plant_water_bucket(
...
+    sm0, drainage0 = soil_plant_water_bucket(
+        sminit,
+        peffective.isel({dim: 0}, drop=True),
+        et.isel({dim: 0}, missing_dims='ignore', drop=True),
+        taw,
+    )
+    # Give time dimension to sm and drainage    
+    sm = sm0.expand_dims(
+        {dim: peffective[dim].size}
+    ).assign_coords({dim: peffective[dim]}).copy()
```
Comment (aaron-kaplan) [inline onset_maproom/agronomy.py:102]: If I understand the docs correctly, you can collapse `expand_dims` and `assign_coords` into a single step:
```
sm = sm0.expand_dims({dim: peffective[dim]})
```

Why did you add the `copy`?

File: onset_maproom/agronomy.py (hunk 1 above)
Comment (aaron-kaplan) [reply onset_maproom/agronomy.py:5]: Will the user of the library need to know call this function, or is it an implementation detail? If an implementation detail, let's make it "private" by either starting its name with an underscore, or adding an `__all__` list and not including this in it.

Comment (remicousin) [reply onset_maproom/agronomy.py:5]: bucket is cute, but is also how they call it, e.g.
https://pressbooks.bccampus.ca/geoglabs2020/wp-content/uploads/sites/1032/2020/08/Figure-8.1-350x292.png
more at:
https://duckduckgo.com/?q=soil+plant+water+balance+bucket+algorithm&t=ffsb&iax=images&ia=images

Comment (remicousin) [reply onset_maproom/agronomy.py:5]: Can we answer the public/private question at the end? I assume private but I don't have a good vision of those things just yet. But yeah I am guessing that at best, we might want to program another kind of step if there are other models than this one, and then we would have some options in the loop function to choose which step-model to use. Or possibly also another loop-function if not compatible.

Comment (aaron-kaplan) [reply onset_maproom/agronomy.py:5]: Oh, then I don't understand why you used this word here. Bucket seems to refer here to a volume of space, whereas this function implements a step in time.

Comment (aaron-kaplan) [reply onset_maproom/agronomy.py:5]: > Can we answer the public/private question at the end? 

Sure.

File: onset_maproom/agronomy.py (hunk 3 above)
Comment (remicousin) [reply onset_maproom/agronomy.py]: if no copy, at some point, when trying to assign other time steps of sm, will get:

ValueError: Assignment destination is a view.  Do you want to .copy() array first?

And that remains true even after collapsing the two functions.

File: onset_maproom/agronomy.py (hunk 1 above)
Comment (remicousin) [reply onset_maproom/agronomy.py]: you do one bucket at a time...! + it's cute indeed. Anyway, it's called _step now

**Review by aaron-kaplan — APPROVED** [review]
